[tuner] Tuner impl.

xwjiang2010 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Would like to get another pass on Tuner Impl. 
In the meantime, I will see to swap the dataset with a lambda to break dependency for easy merging. 

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

krfricke commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,231 @@
+            return
+
+        # Start from fresh
+        assert trainable
```
Not needed now but later we should make sure to validate arguments properly and raise actionable error messages

xwjiang2010 replied:
Acknowledged.

krfricke commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,231 @@
+        local_dir = os.path.dirname(experiment_dir)
+
+        analysis = run(
+            trainable, name=experiment_name, local_dir=local_dir, resume=True
```
The `callbacks` have to be passed again. We will probably have to make a decision here if users have to pass them again or not. Callbacks can have state, so we may have to include them in the experiment checkpoint.
Note that this is currently also not the case in `tune.run` so we don't necessarily have to do it right now.

krfricke replied:
Two things I'm wondering:

1) Should we make "included in checkpoint" the line of what to include in the constructor args vs. the runtime args (i.e. `Tuner(arg=v)` vs `tune.run(arg=val)`? I.e. if callbacks are not checkpointed and thus don't retain their state, make them part of runtime arguments so users don't expect them to retain it on resume?

2) Should we think about a general purpose API for saving/loading checkpoints of arbitrary components, e.g. callbacks? So that we could write checkpointable callbacks without having to include them explicitly in tuner state

krfricke commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,231 @@
+                    # We shouldn't be expecting anything here.
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        if DATASETS in self.param_space:
```
Should we have general support for this in the future, i.e. check all parameters if they are datasets (not just `DATASETS`)? Or will this become unnecessary anyway after ray data fixes?

krfricke commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,231 @@
+                    _helper(v)
+                elif isinstance(v, Dataset):
+                    dataset_dict[k] = v.fully_executed()
+                elif isinstance(v, int):
+                    # CV settings
+                    pass
```
How does this look like in the config?

krfricke commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,231 @@
+            pickle.dump(self.trainable, fp)
+
+    def _process_dataset_param(self):
+        """Some processing on Dataset param."""
```
Let's add an example for a valid config here

krfricke commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,231 @@
+        with open(trainable_ckpt, "wb") as fp:
+            pickle.dump(self.trainable, fp)
+
+    def _process_dataset_param(self):
```
Can we avoid this function altogether? E.g. datasets are fully executed on ser/de, can we just pickle and unpickle the config dict? (or is pickling large datasets large overhead?)

xwjiang2010 replied:
This has nothing to do with ser/deser tho. The datasets passed in are legit dataset objects. It's just that they need to be called with fully_executed() to be passed on to trainables. Updated some of the docstrings to make this clear.

krfricke reviewed (commented):
Looks good so far, have a couple of questions for the datasets part

xwjiang2010 commented:
Hey Kai,
Somehow I cannot reply to some of your comments inline anymore.
So I will address them collectively here.

1. regarding "callbacks" - Changed callbacks to be a "run time" thing. Also made that clear in our docstrings. The implication of this is that user needs to explicitly pass in "callbacks" through Tuner.restore(path, callbacks) if they wish to use them in resume runs. In general, I agree that there is a line to be drawn between things that need to be checkpointed versus not. And for those that do, special care needs to be given to those with states. This is also highly related to how people interact with these APIs, so I propose that we wait till after dogfooding?
2. cv in datasets - I removed it for now. May repurpose it after alpha. Also added examples for how datasets look like.

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+
+
+class Tuner:
+    """The external facing Tuner class as part of unified ML effort.
```
Update docs for something more user-friendly (unified ML is not a user-facing concept)?

xwjiang2010 replied:
Changed to use the wording "Ray MLC"

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+    """The external facing Tuner class as part of unified ML effort.
+
+    Usage pattern:
+    # Only do following, if you want to run in ray client mode.
+    # This means the tune driver code will be running on head node of your cluster.
+    ray.init("ray://my_ray_cluster_address")
+    param_space = {
+        "scaling_config": {
+            "num_actors": tune.grid_search([2, 4]),
+            "cpus_per_actor": 2,
+            "gpus_per_actor": 0,
+        },
+        "preprocessor": tune.grid_search([prep_v1, prep_v2]),
+        "datasets": {
+            "train_dataset": tune.grid_search([ds1, ds2]),
+        },
+        "params": {
+            "objective": "binary:logistic",
+            "tree_method": "approx",
+            "eval_metric": ["logloss", "error"],
+            "eta": tune.loguniform(1e-4, 1e-1),
+            "subsample": tune.uniform(0.5, 1.0),
+            "max_depth": tune.randint(1, 9),
+        },
+    }
+    tuner = Tuner(trainable=trainer, param_space=param_space, name="my_tune_run")
+    analysis = tuner.fit()
```
```suggestion
    .. code-block:: python
    
        # Only do following, if you want to run in ray client mode.
        # This means the tune driver code will be running on head node of your cluster.
        ray.init("ray://my_ray_cluster_address")
        param_space = {
            "scaling_config": {
                "num_actors": tune.grid_search([2, 4]),
                "cpus_per_actor": 2,
                "gpus_per_actor": 0,
            },
            "preprocessor": tune.grid_search([prep_v1, prep_v2]),
            "datasets": {
                "train_dataset": tune.grid_search([ds1, ds2]),
            },
            "params": {
                "objective": "binary:logistic",
                "tree_method": "approx",
                "eval_metric": ["logloss", "error"],
                "eta": tune.loguniform(1e-4, 1e-1),
                "subsample": tune.uniform(0.5, 1.0),
                "max_depth": tune.randint(1, 9),
            },
        }
        tuner = Tuner(trainable=trainer, param_space=param_space, name="my_tune_run")
        analysis = tuner.fit()
```

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+    _local_tuner: Optional[_TunerInternal]  # Only used in none ray client mode.
+    _remote_tuner: Optional[ClientActorHandle]  # Only used in ray client mode.
+
+    def __init__(self, **kwargs):
```
We should document args at this level, not at the TunerInternal level

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+            ).remote(restore_path=path, callbacks=callbacks)
+            return Tuner(tuner_internal=tuner_internal)
+
+    def fit(self):
```
document?

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+                _helper(ds)
+            else:
+                # We shouldn't be expecting anything here.
+                raise TuneError("Unexpected dataset param passed in.")
```
Can we raise a better error message here?

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+        analysis = run(
+            trainable,
+            resume=True,
+            callbacks=self.callbacks,
```
Are all of our callbacks serializable?

xwjiang2010 replied:
User can give whatever they want. So it may not be serializable.
However, this is the same behavior/handling as current `tune.run`

richardliaw reviewed (changes requested):
Generally, the API seems to do the right thing, though I didn't read the datasets and implementation part carefully. 

A couple cosmetic things to state upon first review:
* We need to improve the documentation, making sure args are rendered properly and clearly.
* We need to improve error messages, making sure there are no "assertions" raised or warnings that the user does not understand.
* We should document the whole ray client codepath, as it seems to introduce some complexity here.
* We will add tests, I presume?

I left a couple questions about some edge cases that I can think of,

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+        },
+        """
+
+        def _helper(dataset_dict: dict):
```
@clarkzinzow can you please review this part?

richardliaw commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+        self.__dict__.update(state)
+
+
+class Tuner:
```
Can we tag with @PublicAPI(stability="alpha")?

xwjiang2010 replied:
There is no annotation for alpha?

ericl commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+
+@dataclass
+@PublicAPI
+class TuneAlgoConfig:
```
Move this into a separate file called `tuner_config.py`?

ericl commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+    which is guaranteed to be placed on head node.
+    """
+
+    def __init__(
```
Move this to an impl file.

ericl commented on python/ray/tune/api_v2/dataset_execution_registry.py:
```diff
@@ -0,0 +1,24 @@
+from ray.data import Dataset
```
Do we need an api_v2 folder? It's a bit confusing since Tuner is closely compatible with the original API.

Recommend getting rid of this folder.

ericl commented on python/ray/tune/api_v2/out_of_band_serialize_dataset.py:
```diff
@@ -0,0 +1,34 @@
+import contextlib
```
Consider putting these things in `impl/` sub-dirs.

xwjiang2010 replied:
Done!

xwjiang2010 replied:
> Per @richardliaw 's suggestion, could we split this PR into (1) interface and (2) implementation?
> 
> The interface PR should just be a couple of files with implementation stubbed out (raise NotImplementedError).

Hmmm, yeah I get what you are coming from.
I can see it's beneficial for:
1. When an interface can have several implementations underneath.
2. Special case to 1, where we want to have a dummy impl for unit test purposes.
3. For coordinating and collaboration when concurrent progress needs to be made for various components.

In this Tuner case, I don't think 1 has merit as we only have one real impl. 2 may be relevant especially since Trainer.fit() is calling into Tuner - this could make unit testing Trainer easier. 3. I think we have already passed that stage.

Unique to this Tuner case, is that we have Tuner wraps TunerInternal differently in ray client v.s. non ray client mode. I am concerned that adding another layer and making it Tuner, TunerImpl and TunerInternal may be a little too much.
Could you take a look at the current status of code (as I made quite some updates already, including hiding internal impl details from external users per your suggestion)? If you still think it is better to have this interface abstraction, I am happy to do so!

richardliaw replied:
The merit of splitting interface / implementation here is mainly on the reviewer focus (so, a special case for 3).

In particular, I think Eric's proposal is first to merge in a stub `tuner.py:Tuner` so that we can review/converge on that first, since there are some small minor deviations that are already occuring (RunConfig, etc). After merging those interfaces, we can just follow up with (this PR) to actually fill out the stubs.

richardliaw replied:
In particular, I would propose opening a new PR that merges in stubs + docstrings, and then rebasing this current PR on that.

ericl commented on python/ray/tune/api_v2/convertible_to_trainable.py:
```diff
@@ -0,0 +1,10 @@
+from ray.tune import Trainable
+
+
+class ConvertibleToTrainable(abc.ABC):
```
(PublicAPI annotation)

xwjiang2010 replied:
This is included to Amog's PR. So to be removed actually...

ericl commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+import copy
```
This file has a bunch of classes right now. Could we have it hold only the Tuner class for readability?

xwjiang2010 replied:
split into tuner_internal.py, tuner.py and tuner_config.py

ericl commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+
+
+########################################################################
+# The motivations behind the new Tuner API as part of unified ML effort is that
```
Should remove internal motivation.

xwjiang2010 replied:
are you referring to the consolidated execution layer one?

ericl commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+        restore_path: str = None,
+        trainable: Optional[
+            Union[
+                str,
```
Can we drop str?

xwjiang2010 replied:
hmmm, I personally don't like the idea of overloading too many possibilities in a single argument. However, people do like to use this for RLlib trainables, by, e.g., specifying `PPO`. 
Do you foresee a way that use can still enjoy this conveniency but not at the cost of a clear API?

ericl commented on python/ray/tune/api_v2/tuner.py:
```diff
@@ -0,0 +1,345 @@
+        self._process_dataset_param()
+
+        self.name = name
+        self.callbacks = callbacks
```
Use underscore for private variables.

xwjiang2010 replied:
updated!

ericl reviewed (commented):
Per @richardliaw 's suggestion, could we split this PR into (1) interface and (2) implementation?

The interface PR should just be a couple of files with implementation stubbed out (raise NotImplementedError).

xwjiang2010 commented:
Ready for review again!

amogkam commented on python/ray/tune/tuner.py:
```diff
@@ -0,0 +1,176 @@
+        ] = None,
+        param_space: Optional[Dict[str, Any]] = None,
+        tune_algo_config: Optional[TuneAlgoConfig] = None,
+        name: Optional[str] = None,
+        local_dir: Optional[str] = None,
+        callbacks: Optional[List[Callback]] = None,
```
Are these args supposed to be grouped together as the RunConfig?

richardliaw replied:
+1

xwjiang2010 replied:
Addressed in the API review.

krfricke commented on python/ray/tune/tuner.py:
```diff
@@ -0,0 +1,176 @@
+    tuner = Tuner(trainable=trainer, param_space=param_space, name="my_tune_run")
+    analysis = tuner.fit()
+
+    This returns an ``ExperimentAnalysis`` object, that you can interact with according
+    to its API.
```
Can we maybe just add a shallow wrapper for now? e.g.

```
class ResultGrid:
    def __init__(self, experiment_analysis: ExperimentAnalysis):
        self._experiment_analysis = experiment_analysis

    def _trial_to_result(self, trial: Trial) -> Result:
        result = Result(checkpoint=best_trial.checkpoint, metrics=best_trial.last_result)
        return result

    def get_best_result() -> Result:
        return self._trial_to_result(self._experiment_analysis.best_trial)

    def __len__(self) -> int:
        return len(self._experiment_analysis.trials)
        
    def __getitem__(self, i) -> Result:
        return self._trial_to_result(self.trial[i]) 
```

(untested)

krfricke replied:
(this should probably happen in TunerInternal though)

krfricke commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+            trainable = trainable
+        return trainable
+
+    def fit(self):
```
(and let's move to ResultGrid as discussed below)

maxpumperla commented on python/ray/tune/experiment.py:
```diff
@@ -101,6 +121,18 @@ def __init__(
 
+        local_dir = _get_local_dir_with_expand_user(local_dir)
+        # `experiment_checkpoint_dir` is for internal use only for better
+        # support of Tuner API.
```
shouldn't this go into the doc string for the `Experiment` signature? we should expose document these args, developer API or not.

xwjiang2010 replied:
Updated.

maxpumperla commented on python/ray/tune/experiment.py:
```diff
@@ -283,21 +307,45 @@ def register_if_needed(cls, run_object):
-                    "the type annotations and try again."
-                )
-                raise type(e)(str(e) + " " + extra_msg) from None
+            if not dry_run:
+                try:
+                    register_trainable(name, run_object)
+                except (TypeError, PicklingError) as e:
+                    extra_msg = (
+                        "Other options: "
+                        "\n-Try reproducing the issue by calling "
+                        "`pickle.dumps(trainable)`. "
+                        "\n-If the error is typing-related, try removing "
+                        "the type annotations and try again."
+                    )
+                    raise type(e)(str(e) + " " + extra_msg) from None
```
+1

maxpumperla commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+    The external facing ``Tuner`` multiplexes between local Tuner and remote Tuner
+    depending on whether in Ray client mode.
+
+    In Ray client mode, external Tuner wraps ``_TunerInternal`` into a remote actor,
```
it should say ``TunerInternal``, or is there another one?

maxpumperla commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+        local_dir: Optional[str] = None,
+        callbacks: Optional[List[Callback]] = None,
+    ):
+        """For initialization, there are two scenarios.
```
why not have this in the main doc string and add explanations of all args?

maxpumperla commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+            return
+
+        # Start from fresh
+        assert trainable
```
add error message when this fails

maxpumperla commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+            os.makedirs(path)
+        return path
+
+    # This has to be done through a function signature (@property won't do).
+    def experiment_checkpoint_dir(self):
+        return self.experiment_checkpoint_dir
```
type missing here, too

maxpumperla commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+
+    def _fit_resume(self, trainable):
+        """Fitting for a restored Tuner."""
+        assert self.experiment_checkpoint_dir
```
error message

maxpumperla commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state.pop("trainable", None)
+        state.pop("param_space", None)
```
why pop these?

xwjiang2010 replied:
trainable field is checkpointed separately. 
param_space is not needed for resuming a tune run, as the information is already saved in Searcher and trials.

maxpumperla commented on python/ray/tune/tuner.py:
```diff
@@ -0,0 +1,176 @@
+
+    # Only do the following, if you want to run in ray client mode.
+    # This means the tune driver code will be running on head node of your cluster.
+    ray.init("ray://my_ray_cluster_address")
```
we should start making these doc examples executable. no need to have a pseudo ray cluster address. please add all necessary imports, too. when exposed to docs, this should just run when copied.

maxpumperla commented on python/ray/tune/tuner.py:
```diff
@@ -0,0 +1,176 @@
+            ]
+        ] = None,
+        param_space: Optional[Dict[str, Any]] = None,
+        tune_algo_config: Optional[TuneAlgoConfig] = None,
```
agree should be leaner. but the should we think about renaming `TuneAlgoConfig` too?

xwjiang2010 replied:
Renamed to TuneConfig. Same for the comment below.

maxpumperla commented on python/ray/tune/tuner_config.py:
```diff
@@ -0,0 +1,41 @@
+
+@dataclass
+@PublicAPI(stability="beta")
+class TuneAlgoConfig:
```
or just `TuneConfig`? not sure

maxpumperla commented on python/ray/tune/tuner_config.py:
```diff
@@ -0,0 +1,41 @@
+
+    # Currently this is not at feature parity with `tune.run`, nor should it be.
+    # The goal is to reach a fine balance between API flexibility and conciseness.
+    # We should carefully introduce arguments here instead of just dumping everything.
```
+1, and also think about useful abstractions, instead of just adding flat args to a list.

maxpumperla commented on python/ray/tune/tuner_config.py:
```diff
@@ -0,0 +1,41 @@
+    # We should carefully introduce arguments here instead of just dumping everything.
+    mode: Optional[str] = None
+    metric: Optional[str] = None
+    search_alg: Optional[Searcher] = None
```
Should these come with explicit defaults? Right now I think our `search_alg` default is random search, but I'm not sure we're telling anyone anywhere.

maxpumperla commented on python/ray/tune/tuner.py:
```diff
@@ -0,0 +1,176 @@
+    you can then do
+    .. code-block:: python
+
+    tuner = Tuner.restore(experiment_checkpoint_dir)
```
I'm not entirely convinced we need another checkpoint dir. Are we sure we can't solve this another way? Especially given that this results in yet another argument to `tune.run(...)`.

maxpumperla replied:
also, this example should run. Import `Tuner` and if there's no checkpoint dir, create one.

krfricke replied:
Just as a note, the `experiment_checkpoint_dir` is still the same here, i.e. we only have one dir containing any kind of checkpointed data. Currently `tune.run()` constructs the experiment checkpoint dir from the experiment name and the local dir, and since we use tune.run as the backend we need to be able to set this explicitly and resolve it on the head node (when using Ray client). Hope this makes sense as an explanation.

TLDR: No new directory, just an argument to explicitly overwrite the directory path if needed

ericl commented on python/ray/tune/tuner.py:
```diff
@@ -0,0 +1,176 @@
+# Some requirements on Tuner that may help you understand some implementation decisions:
+# 1. Tuner needs to work in ray client mode.
+# 2. Tuner needs to be restored in case of any failures.
+########################################################################
```
Please clean up the documentation to be suitable for the public audience instead of internal facing.

ericl commented on python/ray/tune/tuner.py:
```diff
@@ -0,0 +1,176 @@
+
+@PublicAPI(stability="beta")
+class Tuner:
+    """The external facing Tuner class as part of Ray MLC effort.
```
Clean up the comment.

ericl reviewed (changes requested):
Discussed with @richardliaw offline and let's try to move the implementations and tests to another PR, so we can focus on reviewing the interfaces first.

(can just put NotImplemented / TODOs for the methods).

xwjiang2010 commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,209 @@
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        if _DATASETS in self.param_space:
+            ds = self.param_space[_DATASETS]
+            if isinstance(ds, Dataset):
+                self.param_space[_DATASETS] = ds.fully_executed()
+            elif isinstance(ds, dict):
+                _helper(ds)
+            else:
+                # We shouldn't be expecting anything here.
+                raise TuneError("Unexpected dataset param passed in.")
```
Hmmm, I cannot think of a general way. Currently, I use "train_dataset" and "extra_datasets" as keys to look for these dataset objects.
I think this is just some contract that needs to be consistent between Tune and Train.

krfricke commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,214 @@
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        for key in _DATASET_KEYS:
+            if key in self._param_space:
+                ds = self._param_space[key]
+                if isinstance(ds, Dataset):
+                    self._param_space[key] = ds.fully_executed()
+                elif isinstance(ds, dict):
+                    _helper(ds)
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
```
Can we just check all config keys?

If I see this correctly it only works with `tune.grid_search` (because it will be converted into a dict) but not with `tune.choice()` (as it will be a `Categorical` object). Can we also support `tune.choice`?

xwjiang2010 commented:
@krfricke I still haven't quite figured out the failed tune multi node test. Rerunning them just to be sure.
I am also going to launch the Tune cloud test suite.

krfricke commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,221 @@
+                elif isinstance(ds, dict):
+                    # grid search is turned into {"grid_search": []}
+                    _helper_dict(ds)
+                elif isinstance(ds, Categorical):
+                    # tune.choice is turned into Categorical object.
+                    for c in ds.categories:
+                        assert isinstance(c, Dataset)
+                        c.fully_executed()
```
This part looks good, but why don't we move this to `helper_dict` and just call `_helper_dict(self._param_space)` (and don't raise TuneErrors)? Then we would preprocess all dataset params (even if a user decides to add more under a different config key) and thus remain forward compatible

krfricke replied:
(We can still validate the default dataset keys but let's make sure that all datasets contained in the config are converted)

xwjiang2010 replied:
Done. Also added a unit test.

krfricke commented:
Is the multinode test blocking this PR?

xwjiang2010 commented:
> Is the multinode test blocking this PR?

Not anymore, the error disappears after merging in master.

krfricke commented on python/ray/tune/impl/utils.py:
```diff
@@ -0,0 +1,29 @@
+            _list = v if isinstance(v, list) else v.categories
+            if len(_list) == 0:
+                return
+            assert all(isinstance(_item, type(_list[0])) for _item in _list)
```
This is a minor nit, but we don't require this constraint in search space definition

krfricke replied:
Let's ignore this for now though

xwjiang2010 replied:
ah good to know! We should add that case to our unit test/examples at some time.

krfricke reviewed (approved):
This looks good to me! Thanks!

Yard1 commented on python/ray/tune/experiment.py:
```diff
@@ -48,6 +49,24 @@ def _validate_log_to_file(log_to_file):
     return stdout_file, stderr_file
 
 
+def _get_local_dir_with_expand_user(local_dir: Optional[str]):
```
```suggestion
def _get_local_dir_with_expand_user(local_dir: Optional[str]) -> str:
```

Yard1 commented on python/ray/tune/experiment.py:
```diff
@@ -48,6 +49,24 @@ def _validate_log_to_file(log_to_file):
+    return os.path.abspath(os.path.expanduser(local_dir or DEFAULT_RESULTS_DIR))
+
+
+def _get_dir_name(run, explicit_name: Optional[str], combined_name: str):
```
```suggestion
def _get_dir_name(run, explicit_name: Optional[str], combined_name: str) -> str:
```

maxpumperla commented on python/ray/tune/experiment.py:
```diff
@@ -73,6 +92,12 @@ class Experiment:
             checkpoint_freq=10,
             max_failures=2)
+
+    Attributes:
```
Args? should be gone with the TODO

xwjiang2010 replied:
We consolidated to using "Attributes" instead of "Args": https://github.com/ray-project/ray/pull/22975#discussion_r824311556

Yard1 commented on python/ray/tune/experiment.py:
```diff
@@ -283,21 +309,67 @@ def register_if_needed(cls, run_object):
+        Assumes already registered if run_object is a string.
+        Also, does not inspect interface of given run_object.
+
+        Arguments:
```
```suggestion
        Args:
```

Yard1 commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,178 @@
+    In Ray client mode, external Tuner wraps ``TunerInternal`` into a remote actor,
+    which is guaranteed to be placed on head node.
+
+    TunerInternal can be constructed from fresh, in which case, `trainable` needs to
+    be provided, together with optional `param_space`, `tune_config` and `run_config`.
+
+    It can also be restored from a previous failed run (given `restore_path`).
```
```suggestion
    TunerInternal can be constructed from fresh, in which case, ``trainable`` needs to
    be provided, together with optional ``param_space``, ``tune_config`` and ``run_config``.

    It can also be restored from a previous failed run (given ``restore_path``).
```

maxpumperla replied:
doc strings, please!

Yard1 commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,178 @@
+    It can also be restored from a previous failed run (given `restore_path`).
+    """
+
+    def __init__(
```
Would be good to add return type hints in all the methods here imo

Yard1 commented on python/ray/tune/tuner.py:
```diff
@@ -1,5 +1,23 @@
+from ray.util.ml_utils.node import force_on_current_node
+
+
+# The magic key that is used when instantiating Tuner during resume.
+_TUNER_INTERNAL = "_tuner_internal"
+_SELF = "self"
```
should those constants be in `ray.ml.constants`?

Yard1 commented on python/ray/tune/tuner.py:
```diff
@@ -1,5 +1,23 @@
+_SELF = "self"
 
 
 @PublicAPI(stability="alpha")
```
Args, not attributes (below)

xwjiang2010 replied:
See https://github.com/ray-project/ray/pull/22975#discussion_r824311556

Yard1 commented on python/ray/tune/tuner.py:
```diff
@@ -96,4 +157,25 @@ def fit(self) -> ResultGrid:
+        if not self._is_ray_client:
+            try:
+                return self._local_tuner.fit()
+            except Exception:
+                raise TuneError(
+                    f"Tune run fails with {traceback.format_exc()}"
+                    f'Please use tuner = Tuner.restore("'
+                    f'{self._local_tuner.experiment_checkpoint_dir}") to resume.'
+                )
```
```suggestion
            except Exception as e:
                raise TuneError(
                    f"Tune run failed."
                    f'Please use tuner = Tuner.restore("'
                    f'{self._local_tuner.experiment_checkpoint_dir}") to resume.'
                ) from e
```

Yard1 commented on python/ray/tune/tuner.py:
```diff
@@ -96,4 +157,25 @@ def fit(self) -> ResultGrid:
+            )
+            try:
+                return ray.get(self._remote_tuner.fit.remote())
+            except Exception:
+                raise TuneError(
+                    f"Tune run fails with {traceback.format_exc()}"
+                    f'Please use tuner = Tuner.restore("'
+                    f'{experiment_checkpoint_dir}") to resume.'
+                )
```
```suggestion
            except Exception as e:
                raise TuneError(
                    f"Tune run failed."
                    f'Please use tuner = Tuner.restore("'
                    f'{experiment_checkpoint_dir}") to resume.'
                ) from e
```

Yard1 reviewed (commented):
Looks good! Some nits

maxpumperla commented on python/ray/tune/impl/tuner_internal.py:
```diff
@@ -0,0 +1,178 @@
+            return
+
+        # Start from fresh
+        if not trainable:
```
why is is optional then?

xwjiang2010 replied:
ah, if "restore_path" is given, then all the rest is not needed, including trainable, thus optional.

amogkam commented on python/ray/tune/result_grid.py:
```diff
@@ -62,7 +62,6 @@ def _populate_exception(trial: Trial) -> Optional[TuneError]:
-        # TODO(xwjiang): Use Kai's new checkpoint!
         result = Result(
             checkpoint=trial.checkpoint,
             metrics=trial.last_result,
```
Is it possible to get back the entire sequence of reported metrics, not just the last result?

richardliaw commented on python/ray/tune/impl/test_utils.py:
```diff
@@ -0,0 +1,58 @@
+    executed_ds = param_space["train_dataset"].categories
+    assert len(executed_ds) == 2
+    assert executed_ds[0]._plan._has_final_stage_snapshot()
+    assert executed_ds[1]._plan._has_final_stage_snapshot()
```
I think you need a __main__ on this file so that it actually runs

xwjiang2010 replied:
Added main. The test won't run for now tho. Will add the build rule later after datasets is landed.
